### PR TITLE
chore: fix flaky tests on CI

### PR DESCRIPTION
**Pull request by alessbell**
This PR fixes some existing issues with our current test suite in at attempt address the flakiness that has been an issue for the past few months.

I've tried to limit its scope, but as you can see from the commit history there was lots of trial and error 😄 

This PR:
- adds `eslint` and `eslint-plugin-testing-library`, runs it on CCI against all test files and fixes errors
- adds some missing `mockRestore()` calls
- adds `IS_REACT_18` conditions around assertions in `src/react/hooks/tests/useLazyQuery.test.tsx` and `src/react/hooks/tests/useQuery.test.tsx`

It would seem that once we began following the Testing Library rules strictly, as enforced by the ESLint plugin, some tests (`useLazyQuery - 3 instances`, `useQuery - 13 instances`) began consistently failing when run against React 17. For now, I adjusted the tests to assert something different based on React version as we do elsewhere, but feel free to take a closer look and/or I can create a separate ticket to investigate those.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -29,25 +29,6 @@ jobs:
       - store_artifacts:
           path: reports/junit
 
-  # Ensure that any PR that changes packages has a changeset on it (perhaps
```
Comment (alessbell) [inline .circleci/config.yml:32]: This wasn't being run, and `unless` appears to be an old CCI syntax so I'm removing for now.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (alessbell) [inline .eslintrc:1]: Adds ESLint rules for our tests to prevent future memory leaks/bugs 🎉

File: .circleci/config.yml [hunk 3]
```diff
@@ -3,17 +3,25 @@ version: 2
...
       - run: npm run bundlesize
 
+  Lint:
+    docker:
+      - image: cimg/node:18.12.0
+    steps:
+      - checkout
+      - run: npm version
+      - run: npm ci
+      - run: npm run lint
```
Comment (alessbell) [inline .circleci/config.yml:20]: Adding a `Lint` job to run eslint on test files only.

File: package.json [hunk 4]
```diff
@@ -44,13 +44,14 @@
...
     "test": "jest --config ./config/jest.config.js",
-    "test:debug": "node --inspect-brk node_modules/.bin/jest --config ./config/jest.config.js --runInBand --testTimeout 99999",
-    "test:ci": "TEST_ENV=ci npm run test:coverage && npm run test:memory",
+    "test:debug": "node --inspect-brk node_modules/.bin/jest --config ./config/jest.config.js --runInBand --testTimeout 99999 --logHeapUsage",
+    "test:ci": "TEST_ENV=ci npm run test:coverage -- --logHeapUsage && npm run test:memory",
     "test:watch": "jest --config ./config/jest.config.js --watch",
     "test:memory": "cd scripts/memory && npm i && npm test",
     "test:coverage": "npm run coverage -- --ci --runInBand --reporters=default --reporters=jest-junit",
-    "coverage": "jest --config ./config/jest.config.js --verbose --coverage",
+    "coverage": "jest --config ./config/jest.config.js --coverage",
```
Comment (alessbell) [inline package.json:54]: I personally found the verbose output very noisy on CI (but happy to put this back if folks want!)

File: package.json [hunk 5]
```diff
@@ -44,13 +44,14 @@
     "postprocess-dist": "ts-node-script config/postprocessDist.ts",
     "clean": "rimraf -r dist coverage lib temp",
     "ci:precheck": "node config/precheck.js",
+    "lint": "eslint '**/__tests__/**/*.[jt]sx'",
     "test": "jest --config ./config/jest.config.js",
-    "test:debug": "node --inspect-brk node_modules/.bin/jest --config ./config/jest.config.js --runInBand --testTimeout 99999",
-    "test:ci": "TEST_ENV=ci npm run test:coverage && npm run test:memory",
+    "test:debug": "node --inspect-brk node_modules/.bin/jest --config ./config/jest.config.js --runInBand --testTimeout 99999 --logHeapUsage",
+    "test:ci": "TEST_ENV=ci npm run test:coverage -- --logHeapUsage && npm run test:memory",
```
Comment (alessbell) [inline package.json:50]: On ci, we now log heap usage inline with ever test that executes. This should help us investigate flakiness in the future.

**Comment by alessbell** [comment]
Also fwiw, I tried removing the `--runInBand` flag on CI but it produced some [interesting failures](https://app.circleci.com/pipelines/github/apollographql/apollo-client/18189/workflows/b2fb10da-67c7-4e00-8183-7c9b46425ff2/jobs/96415) (`A jest worker process (pid=428) was terminated by another process: signal=SIGKILL, exitCode=null. Operating system logs may contain more information on why this occurred.`)

On my M1 mac, however, I can run `npm run test` and have all tests pass in ~25s:

CleanShot 2023-01-11 at 15 19 27

File: .eslintrc (hunk 2 above)
Comment (jerelmiller) [reply .eslintrc:1]: This is great! I'd love to expand this in the future for other things as well (no unused vars, etc). ESLint is such a valuable tool. Thanks for starting with tests!

File: src/react/components/__tests__/client/Mutation.test.tsx [hunk 6]
```diff
@@ -1420,7 +1433,7 @@ describe('General Mutation testing', () => {
     console.log = errorLogger;
   });
 
-  itAsync('errors when changing from mutation to a query', (resolve, reject) => {
+  it('errors when changing from mutation to a query', async () => {
```
Comment (jerelmiller) [inline src/react/components/__tests__/client/Mutation.test.tsx:1436]: 🎉

File: src/react/hoc/__tests__/queries/errors.test.tsx [hunk 7]
```diff
@@ -220,35 +221,30 @@ describe('[queries] errors', () => {
...
+                // the second request had an error!
+                expect(props.data!.error).toBeTruthy();
+                expect(props.data!.error!.networkError).toBeTruthy();
+                // // We need to set a timeout to ensure the unhandled rejection is swept up
+                setTimeout(() => {
+                  // try {
+                  expect(unhandled.length).toEqual(0);
+                  // } catch (err) {
+                  //   reject(err);
+                  // }
```
Comment (jerelmiller) [inline src/react/hoc/__tests__/queries/errors.test.tsx:245]:
```suggestion
                  expect(unhandled.length).toEqual(0);
```

Out of curiosity, any reason we couldn't just remove this?

File: src/react/hoc/__tests__/queries/observableQuery.test.tsx [hunk 8]
```diff
@@ -57,7 +56,7 @@ describe('[queries] observableQuery', () => {
       options: { fetchPolicy: 'cache-and-network' }
     })(
       class extends React.Component<ChildProps<{}, Data>> {
-        componentDidUpdate() {
+        async componentDidUpdate() {
```
Comment (jerelmiller) [inline src/react/hoc/__tests__/queries/observableQuery.test.tsx:59]: [nit] I don't see an `await` keyword added in here. Should the `userEvent.click` on line 68 have an `await` in front of it? If not, you should be able to remove this keyword.

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 9]
```diff
@@ -4982,68 +5091,132 @@ describe('useQuery Hook', () => {
...
+        const { data } = result.current.useQueryResult;
         expect(data).toEqual({ a: "aa", b: 1 });
+      }, { interval: 1 });
+      await waitFor(() => {
+        const { previousData } = result.current.useQueryResult;
         expect(previousData).toEqual({ a: "a" });
       }, { interval: 1 });
 
-      await expect(waitFor(() => {
+      await expect(await waitFor(() => {
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useQuery.test.tsx:5162]: [nit] Does this need the `expect` wrapped around this `waitFor`?

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 10]
```diff
@@ -4982,68 +5091,132 @@ describe('useQuery Hook', () => {
...
+        const { data } = result.current.useQueryResult;
         expect(data).toEqual({ a: "aaa", b: 2 });
+      }, { interval: 1 });
+      await waitFor(() => {
+        const { previousData } = result.current.useQueryResult;
         expect(previousData).toEqual({ a: "aa", b: 1 });
       }, { interval: 1 });
 
-      await expect(waitFor(() => {
+      await expect(await waitFor(() => {
```
Comment (jerelmiller) [inline src/react/hooks/__tests__/useQuery.test.tsx:5182]: [nit] Does this need the expect wrapped around this waitFor?

**Review by jerelmiller — APPROVED** [review]
Thanks so much for handling this! I know this kind of grunt work isn't super fun sometimes, but this is very valuable. Feel free to address or ignore my [nit] comments as you see fit.

File: src/react/hooks/__tests__/useQuery.test.tsx (hunk 9 above)
Comment (alessbell) [reply src/react/hooks/__tests__/useQuery.test.tsx:5162]: Thanks for catching these!

File: src/react/hoc/__tests__/queries/errors.test.tsx (hunk 7 above)
Comment (alessbell) [reply src/react/hoc/__tests__/queries/errors.test.tsx:245]: Oops, forgot to remove - thanks!

**Pull request merged**